tempesta_fw/sock.c

```diff
@@ -483,7 +551,7 @@ __ss_close(struct sock *sk, int flags)
 	ss_do_close(sk);
 	bh_unlock_sock(sk);
 	if (flags & SS_F_CONN_CLOSE)
-		SS_CALL(connection_drop, sk);
+		SS_CALL_GUARD_EXIT(connection_drop, sk);
```

Comment:
I have spent quite some time trying to think of different scenarios that may occur here. This is tricky.

So, the case of asynchronous close() is covered by a check for `SOCK_DEAD`. A close() is asynchronous unless a specific set of criteria are satisfied, and one criteria in the set is that we're in SoftIRQ. 

In the case of synchronous close() [this should really be called *immediate* here as `ss_close_sync()` may enqueue a close() the same way as an asynchronous `ss_close()`, and so it would behave the same way as an asynchronous close()] the point is that all calls into SS/Tempesta are serialized with a socket lock. SS is not called by the kernel when the socket is dead (SOCK_DEAD). It means that if we're here in SS and we're doing a synchronous/immediate close(), then the socket is alive at this moment.

Is that correct? Perhaps it's best to document this.

Unless somehow an immediate socket close() is called twice in succession by Tempesta in one SoftIRQ, we're safe.

Can it be called twice? I am not sure yet. Perhaps, that must be construed as a bug. Right now, there's no protection against that, so it might be a good idea to put a trap for it.